src/hotspot/share/gc/g1/g1UncommitRegionTask.cpp

```diff
@@ -0,0 +1,135 @@
+  _instance->set_active(true);
+  G1CollectedHeap::heap()->service_thread()->register_task(_instance);
+}
+
+G1UncommitRegionTask* G1UncommitRegionTask::instance() {
+  if (_instance == NULL) {
+    initialize();
+  }
+  return _instance;
+}
+
+void G1UncommitRegionTask::run() {
+  assert_at_safepoint_on_vm_thread();
+
+  G1UncommitRegionTask* uncommit_task = instance();
+  if (!uncommit_task->is_active()) {
+    // Change state to active and schedule with no delay.
+    uncommit_task->set_active(true);
+    uncommit_task->schedule(0);
+    // Notify service thread to avoid unnecessary waiting.
+    G1CollectedHeap::heap()->service_thread()->notify();
+  }
+}
+
+bool G1UncommitRegionTask::is_active() {
+  return _active;
+}
+
+void G1UncommitRegionTask::set_active(bool state) {
+  assert(_active != state, "Must do a state change");
+  _active = state;
+}
+
+void G1UncommitRegionTask::report_execution(Tickspan time, uint regions) {
+  _summary_region_count += regions;
+  _summary_duration += time;
+
+  log_trace(gc, heap)("Concurrent Uncommit: " SIZE_FORMAT "%s, %u regions, %1.3fms",
+                      byte_size_in_proper_unit(regions * HeapRegion::GrainBytes),
+                      proper_unit_for_byte_size(regions * HeapRegion::GrainBytes),
+                      regions,
+                      time.seconds() * 1000);
+}
+
+void G1UncommitRegionTask::report_summary() {
+  log_debug(gc, heap)("Concurrent Uncommit Summary: " SIZE_FORMAT "%s, %u regions, %1.3fms",
+                      byte_size_in_proper_unit(_summary_region_count * HeapRegion::GrainBytes),
+                      proper_unit_for_byte_size(_summary_region_count * HeapRegion::GrainBytes),
+                      _summary_region_count,
+                      _summary_duration.seconds() * 1000);
+}
+
+void G1UncommitRegionTask::clear_summary() {
+  _summary_duration = Tickspan();
+  _summary_region_count = 0;
+}
+
+void G1UncommitRegionTask::execute() {
+  assert(_active, "Must be active");
+
+  // Each execution is limited to uncommit at most 256M worth of regions. This
+  // limit is small enough to ensure that the duration of each invocation is
+  // short, while still making reasonable progress.
+  static const uint region_limit = (uint) (256 * M / G1HeapRegionSize);
+
+  // Prevent from running during a GC pause.
+  SuspendibleThreadSetJoiner sts;
+  G1CollectedHeap* g1h = G1CollectedHeap::heap();
+
+  Ticks start = Ticks::now();
+  uint uncommit_count = g1h->uncommit_regions(region_limit);
+  Tickspan uncommit_time = (Ticks::now() - start);
+
+  if (uncommit_count > 0) {
+    report_execution(uncommit_time, uncommit_count);
+  }
+
+  // Reschedule if there are more regions to uncommit, otherwise
+  // change state to inactive.
+  if (g1h->has_uncommittable_regions()) {
+    // No delay, reason to reschedule rather then to loop is to allow
+    // other tasks to run without waiting for a full uncommit cycle.
+    schedule(0);
```

Comment:
You are correct that doing the extra notification won't hurt, but I don't really see why we should do the notification when we know it is not needed. There are probably a few improvements that we want to do when it comes to how the service thread is registering and scheduling tasks. Since this mechanism is very new I think we will realize how we want this to work more and more.

This is the way I see it right now (after the introduction of the public `G1ServiceThread::schedule_task()`:
* `G1ServiceTask::schedule(delay)` should only be called from a running task. Using it to schedule it from the outside was a bit of a hack. `schedule()` will call `G1ServiceThread::schedule(task, delay)` and nothing more.
* `G1ServiceThread::schedule_task(task, delay)` should be used to schedule a task from the outside. It will under the hood call `G1ServiceThread::schedule(task, delay)` and `G1ServiceThread::notify()` to handle the case where the task end up being the first task in the queue.

To me separating the use cases is good, but you might not agree.  We could add an assert to `G1ServiceTask::schedule(delay)` to ensure that it is only called when running on the service thread, that way we would catch wrong usage of the API quickly.

Reply:
I would be fine with the assert to prevent misuse.

Reply:
Just pushed the assert and also update the comments a bit.